### PR TITLE
factors out controller context and moves synchronisation logic

### PR DESCRIPTION
**Pull request by p0lyn0mial**
**What this PR does / why we need it**: the idea is that we will create the controller from the context in a "loop". In addition this pull also moves synchronisation logic down to the controller itself.

**Which issue(s) this PR fixes**: see #58

### CODE REVIEW

**Comment by p0lyn0mial** [comment]
we can hold this pull, I'll issue a few more so that we can see where this all is going and we can decide if we like it or not.

**Comment by mrIncompetent** [comment]
/assign @metalmatze 
I'm a bit swamped right now

File: pkg/controller/machine.go [hunk 1]
```diff
@@ -94,15 +95,27 @@ type MetricsCollection struct {
...
+	metrics MetricsCollection,
+	stopCh <-chan struct{}) *Controller {
+
+	go kubeInformerFactory.Start(stopCh)
+	go machineInformerFactory.Start(stopCh)
+
+	for _, syncsMap := range []map[reflect.Type]bool{kubeInformerFactory.WaitForCacheSync(stopCh), machineInformerFactory.WaitForCacheSync(stopCh)} {
+		for key, synced := range syncsMap {
+			if !synced {
+				glog.Fatalf("unable to sync %s", key)
```
Comment (p0lyn0mial) [inline pkg/controller/machine.go]: I would prefer to return an error :)

**Comment by mrIncompetent** [comment]
Very important thing: please rebase first. The arguments for NewController changed.
We stopped passing in the informerFactories and instead passed in, the required informers

**Comment by p0lyn0mial** [comment]
> Very important thing: please rebase first. The arguments for NewController changed.

grr.

File: cmd/controller/main.go [hunk 2]
```diff
@@ -50,6 +50,63 @@ var (
...
+
+	// TODO: desc
+	machineInformerFactory machineinformers.SharedInformerFactory
+
+	// TODO: desc
+	stopCh <-chan struct{}
+}
+
+// TODO: desc
+func createClientsOrDie(cfg *rest.Config) (kubernetes.Interface, apiextclient.Interface, machineclientset.Interface) {
```
Comment (metalmatze) [inline cmd/controller/main.go]: I think it's fine to create these clients simply in the main func. We don't need to wrap these things.
Plus: Do we really want to return interfaces? You should always return concrete types. Other funcs can then ask for Interfaces.

File: cmd/controller/main.go [hunk 3]
```diff
@@ -50,6 +50,63 @@ var (
 	workerCount   int
 )
 
+// TODO:desc
+type controllerContext struct {
```
Comment (metalmatze) [inline cmd/controller/main.go]: It's weird calling this context. Not really sure if that's a common thing within Kubernetes, but these days, when I see something called context I mostly expect it to use the `context` package. :thinking:

File: pkg/controller/machine.go [hunk 4]
```diff
@@ -94,15 +95,27 @@ type MetricsCollection struct {
 	NodeJoinDuration    metrics.Histogram
 }
 
-// NewMachineController returns a new machine controller
-func NewMachineController(
+// NewMachineControllerOrDie returns a new machine controller
+func NewMachineControllerOrDie(
```
Comment (metalmatze) [inline pkg/controller/machine.go]: Why don't we simply make this a `func NewMachineController(...) (*Controller, error) {` and let the caller decide what they want to do?
The caller can still decide if they want to fail (log something and then exit != 0) or if they want to maybe backoff and retry, or simply continue.

Comment (p0lyn0mial) [reply pkg/controller/machine.go]: yes, good idea.

File: cmd/controller/main.go (hunk 3 above)
Comment (p0lyn0mial) [reply cmd/controller/main.go]: sure we can change the name, I don't have strong naming preferences.

**Pull request closed without merging**